src/system/SystemLayerImplSelect.h

```diff
@@ -29,6 +29,13 @@
 #include <pthread.h>
 #endif // CHIP_SYSTEM_CONFIG_POSIX_LOCKING
 
+#if CHIP_SYSTEM_CONFIG_USE_LIBEV
+#include <ev.h>
+#if CHIP_SYSTEM_CONFIG_USE_DISPATCH
+#error "CHIP_SYSTEM_CONFIG_USE_LIBEV and CHIP_SYSTEM_CONFIG_USE_DISPATCH are mutually exclusive"
```

Comment:
Right, so they are mutually exclusive, so why the platform manager changes on Darwin?

The Darwin platform manager _always_ uses dispatch, period.

Reply:
> The Darwin platform manager _always_ uses dispatch, period.

Not necessarily ;-) Using dispatch *for I/O and timers* is a decision orthogonal to using the Darwin platform manager. In fact, for *development* of my libev based app (on the mac, in XCode, just because it's so much more comfortable to use…), I use the Darwin platform, but just swap dispatch for libev.
Of course, this is never going to be a production app configuration, so I'd understand you wouldn't want it in mainline. On the other hand, these types of unorthodox setups tend to reveal otherwise hard to find platform dependencies and edge cases.

Reply:
> I use the Darwin platform, but just swap dispatch for libev.

How do you guarantee that I/O processing is happening on the right dispatch queue in that case, and in particular that it does not race with other Matter things that do run on that dispatch queue?

Reply:
> How do you guarantee that I/O processing is happening on the right dispatch queue in that case

Without CHIP_SYSTEM_CONFIG_USE_DISPATCH, nothing in the non-Darwin-specific code can run anywhere else than under libev control (timer and IO callbacks) on the same thread/queue.

Only Darwin specific code *could* indirectly make use of APIs that would call back from other queue/thread contexts. Of those, my already-on-network app only uses dns-sd and only for publishing. I cannot *guarantee* anything here, I just tried it works in my context, which is, as said, **not** a setup for a productive build, but just a development convenience. For that, the possible dangers are isolated enough.

Bottom line - I'd understand if you didn't want `#if`s in the platform manager that only support a non-100% safe development setup. You decide 😄

Reply:
> Without CHIP_SYSTEM_CONFIG_USE_DISPATCH, nothing in the non-Darwin-specific code can run anywhere else than under libev control (timer and IO callbacks) on the same thread/queue.

Does this assume that something other than src/platform/darwin/PlatformManagerImpl.cpp is being used as the platform manager?  There's a not-entirely-obvious dependency here where ScheduleLambda on the System::Layer will run it via PostEvent on the platform manager...

And of course various things directly PostEvent and whatnot.

If you're not using the platform manager's event loop at all (so not calling RunEventLoop or StartEventLoopTask), then I'd like to understand the setup and what things are running.  If you _are_, then the assumption is that PostEvent and whatever System::Layer is doing run things on the same thread, or at least interlocked sanely so they can't run at the same time.

Mostly I feel that event loops and the stuff around them are complex enough people get confused easily, so I would rather not have code around that makes them _more_ confused.